src/core/lib/event_engine/windows/windows_engine.cc

```diff
@@ -253,38 +349,60 @@ bool WindowsEventEngine::IsWorkerThread() { grpc_core::Crash("unimplemented"); }
 void WindowsEventEngine::OnConnectCompleted(
     std::shared_ptr<ConnectionState> state) {
   absl::StatusOr<std::unique_ptr<WindowsEndpoint>> endpoint;
-  EventEngine::OnConnectCallback cb;
   {
     // Connection attempt complete!
-    grpc_core::MutexLock lock(&state->mu);
-    cb = std::move(state->on_connected_user_callback);
-    state->on_connected_user_callback = CreateCrashingOnConnectCallback();
-    state->on_connected = nullptr;
+    grpc_core::MutexLock lock(&state->mu());
+    // return early if we cannot cancel the connection timeout timer.
+    int erased_handles = 0;
     {
       grpc_core::MutexLock handle_lock(&connection_mu_);
-      known_connection_handles_.erase(state->connection_handle);
+      erased_handles =
+          known_connection_handles_.erase(state->connection_handle());
     }
-    const auto& overlapped_result = state->socket->write_info()->result();
-    // return early if we cannot cancel the connection timeout timer.
-    if (!Cancel(state->timer_handle)) return;
+    if (erased_handles != 1 || !Cancel(state->timer_handle())) {
+      GRPC_EVENT_ENGINE_TRACE(
+          "%s", "Not accepting connection since the deadline timer has fired");
+      return;
+    }
+    // Release refs held by the deadline timer.
+    state->AbortDeadlineTimer();
+    const auto& overlapped_result = state->socket()->write_info()->result();
     if (!overlapped_result.error_status.ok()) {
-      state->socket->Shutdown(DEBUG_LOCATION, "ConnectEx failure");
+      state->socket()->Shutdown(DEBUG_LOCATION, "ConnectEx failure");
       endpoint = overlapped_result.error_status;
     } else if (overlapped_result.wsa_error != 0) {
-      state->socket->Shutdown(DEBUG_LOCATION, "ConnectEx failure");
+      state->socket()->Shutdown(DEBUG_LOCATION, "ConnectEx failure");
       endpoint = GRPC_WSA_ERROR(overlapped_result.wsa_error, "ConnectEx");
     } else {
-      ChannelArgsEndpointConfig cfg;
-      endpoint = std::make_unique<WindowsEndpoint>(
-          state->address, std::move(state->socket), std::move(state->allocator),
-          cfg, thread_pool_.get(), shared_from_this());
+      endpoint = state->FinishConnectingAndMakeEndpoint(thread_pool_.get());
     }
   }
   // This code should be running in a thread pool thread already, so the
   // callback can be run directly.
+  auto cb = state->TakeCallback();
+  state.reset();
   cb(std::move(endpoint));
 }
 
+void WindowsEventEngine::OnDeadlineTimerFired(
+    std::shared_ptr<ConnectionState> connection_state) {
+  bool cancelled = false;
+  {
+    grpc_core::MutexLock lock(&connection_state->mu());
+    cancelled = CancelConnectFromDeadlineTimer(connection_state.get());
+  }
+  if (cancelled) {
+    auto cb = connection_state->TakeCallback();
+    connection_state.reset();
+    cb(absl::DeadlineExceededError("Connection timed out"));
+    return;
+  }
+  GRPC_EVENT_ENGINE_TRACE(
```

Comment:
Nit: I dont think this needs to be traced.

Reply:
Trace removed.